Docs/docker compose update (#64565)
grafana/grafana: docs/sources/setup-grafana/configure-docker.md

  - configuration
  - documentation
  - docker
  - docker compose

> title: Configure Docker image

Reviewer: ```suggestion
title: Configure a Grafana Docker image
```

weight: 1800
---

# Configure a Grafana Docker image

> Running Grafana on Docker may require some additional configuration which depends on the complexity of your environment. We will explain some of those e.g. using different images, changing logging levels, defining secrets on Cloud (e.g. AWS), configuring plugins etc.

Reviewer: ```suggestion
Running Grafana on Docker might require some additional configuration depending on your environment's complexity. We will explain some of those e.g. using different images, changing logging levels, defining secrets on Cloud (e.g. AWS), configuring plugins etc.
```

Reviewer: what do you mean by "explain some of those"? Explain some of what?

Also, we want to avoid use of the word 'we'. You can simply say "This topic includes/explains etc."

Author: I cannot explain the "some of those" as already defined as an e.g. in the text i.e. e.g. using different images, changing logging levels, defining secrets on Cloud (e.g. AWS), configuring plugins etc.

Yes, agree to remove "we" from the above.

We can talk more on this one via chat

Reviewer: Ok, I updated, please take a look.

> **Note:** These examples use the Grafana Enterprise docker image. You can use the Grafana Open Source edition by changing the docker image to `grafana/grafana-oss`.

> # Supported Docker Image variants

Reviewer: ```suggestion
# Supported Docker image variants
```

Reviewer: We use sentence case capitalization, not title case.

You can install and run Grafana using the following official Docker images.

> **Grafana Enterprise**: `grafana/grafana-enterprise`

Reviewer: ```suggestion
- **Grafana Enterprise**: `grafana/grafana-enterprise`
```

> **Grafana Open Source**: `grafana/grafana-oss`

Reviewer: ```suggestion
- **Grafana Open Source**: `grafana/grafana-oss`
```

> Each edition is available in two variants: Alpine and Ubuntu. See below.

Reviewer: ```suggestion
Each edition is available in two variants: Alpine and Ubuntu.
```

## Alpine image (recommended)

> The default images are based on the popular [Alpine Linux project](http://alpinelinux.org/), available in the [Alpine official image](https://hub.docker.com/_/alpine). Alpine Linux is much smaller than most distribution base images, and thus leads to slimmer and more secure images.

Reviewer: ```suggestion
By default, the images are built using the widely-used [Alpine Linux project](http://alpinelinux.org/) and can be accessed through the [Alpine official image](https://hub.docker.com/_/alpine). Alpine Linux is much smaller than other distribution base images, allowing for slimmer and more secure images to be created.
```

> The Alpine variant is highly recommended when security is important and you want an  image size that’s as small as possible. Note that the Alpine variant uses [musl libc](http://www.musl-libc.org/) instead of [glibc and friends](http://www.etalabs.net/compare_libcs.html), so certain software might run into issues depending on the depth of their libc requirements. However, most software doesn’t have an issue with this, so this variant is usually a very safe choice.

Reviewer: ```suggestion
If you prioritize security and want to minimize the size of your image, it is recommended that you use the Alpine variant. However, it's important to note that the Alpine variant uses [musl libc](http://www.musl-libc.org/) instead of [glibc and others](http://www.etalabs.net/compare_libcs.html). As a result, some software might encounter problems depending on their libc requirements. Nonetheless, most software should not experience any issues, so the Alpine variant is generally reliable.
```

## Ubuntu image

> 

Reviewer: Need introductory text here before the unordered list.

Author: done please check

Reviewer: I don't see an intro.

> **Grafana Enterprise**: `grafana/grafana-enterprise:<version>-ubuntu`

Reviewer: ```suggestion
- **Grafana Enterprise**: `grafana/grafana-enterprise:<version>-ubuntu`
```

> **Grafana Open Source**: `grafana/grafana-oss:<version>-ubuntu`

Reviewer: ```suggestion
- **Grafana Open Source**: `grafana/grafana-oss:<version>-ubuntu`
```

> These images are based on [Ubuntu](https://ubuntu.com/), available in the [Ubuntu official image](https://hub.docker.com/_/ubuntu). It is an alternative image for those who prefer an Ubuntu-based image and/or are dependent on certain tooling not available for Alpine.

Reviewer: ```suggestion
These images are based on [Ubuntu](https://ubuntu.com/) and can be accessed through the [Ubuntu official image](https://hub.docker.com/_/ubuntu). This is a good option for users who prefer an Ubuntu-based image or require specific tools unavailable on Alpine.
```

Author: These images is reffering to :
Grafana Enterprise: grafana/grafana-enterprise:<version>-ubuntu
and
Grafana Open Source: grafana/grafana-oss:<version>-ubuntu

Reviewer: Got it. Modified my suggestion. Feel free to accept if it looks good to you.

## Run a specific version of Grafana

You can also run a specific version, or run a beta version based on the main branch of the [grafana/grafana GitHub repository](https://github.com/grafana/grafana).

> > **Note:** If you are on a Linux system (e.g. Debian, Ubuntu), you might need to add `sudo` before the command or add your user to the `docker` group by following the official documentation to [run Docker without a non-root user](https://docs.docker.com/engine/install/linux-postinstall/).

Reviewer: ```suggestion
> **Note:** If you use a Linux operating system such as Debian or Ubuntu and encounter permission errors when running Docker commands, you might need to prefix the command with `sudo` or add your user to the `docker` group. The official Docker documentation provides instructions on how to [run Docker without a non-root user](https://docs.docker.com/engine/install/linux-postinstall/).

```

To run a specific version of Grafana mention it in the command <version number> section:

```bash
docker run -d -p 3000:3000 --name grafana grafana/grafana-enterprise:<version number>
```

Example:

> This command will run Grafana Enterprise container which has version 9.4.7. If you want to run a different version e.g. 9.4.3 then just put it in the version-number section

Reviewer: ```suggestion
The following command runs the Grafana Enterprise container, which has version 9.4.7. If you want to run a different version, modify the version-number section.
```

```bash
docker run -d -p 3000:3000 --name grafana grafana/grafana-enterprise:9.4.7
```

## Run the Grafana main branch

> For every successful build of the main branch, we update the `grafana/grafana-oss:main` and `grafana/grafana-oss:main-ubuntu` tags. Additionally, two new tags are created, `grafana/grafana-oss-dev:<version>-<build ID>pre` and `grafana/grafana-oss-dev:<version>-<build ID>pre-ubuntu`, where version is the next version of Grafana and build ID is the ID of the corresponding CI build. You can use these Tags to get access to the latest main builds of Grafana.

Reviewer: ```suggestion
After every successful build of the main branch, two tags, `grafana/grafana-oss:main` and `grafana/grafana-oss:main-ubuntu`, are updated. Additionally, two new tags are created: `grafana/grafana-oss-dev:-pre` and `grafana/grafana-oss-dev:-pre-ubuntu`, where version is the next version of Grafana and build ID is the ID of the corresponding CI build. You can use these tags to get access to the latest main builds of Grafana.
```

Reviewer: I don't quite understand what "where version is the next version of Grafana and build ID is the ID of the corresponding CI build." means - I don't see version or build ID in the tags.

Author: These need to be rechecked by a developer as it is taken from the original (aka current documentation) and I did not do any modification. To what I understand that it is about the docker images available on
https://hub.docker.com/layers/grafana/grafana-oss/main-ubuntu/images/sha256-d42e791118547736a3bea01adb978055bad6d0cfee63914e37737ff1e12f9d8d

and 

https://hub.docker.com/layers/grafana/grafana-oss/main/images/sha256-43b0cf88ad64b3947d0f41654070ae0189b7a44674abbdcd22052053b28e1c19?context=explore

Dan might also help here.

Reviewer: Acknowledged, thanks.

Reviewer 2: `<version>` and `<build ID>` are placeholders, if you look at the docker repo you can see the tags like [10.1.0-55763pre](https://hub.docker.com/layers/grafana/grafana-enterprise-dev/10.1.0-55763pre/images/sha256-b13066ff05063e3b7e4b9128ce93398858711b068abde717de1df2f37d45eb0b?context=explore) (where version is 10.1.0 and build id is 55763)

The main and main-ubuntu tags in the main grafana-oss repo are used to run the latest build from the main branch, the specific tags in the grafana-oss-dev repo are used if you want to run unreleased versions but control which specific build you want.  Neither of these are really useful to a typical end user though as in both cases you're running on the bleeding edge.

> When running Grafana main in production, we strongly recommend that you use the `grafana/grafana-oss-dev:<version>-<build ID>pre` tag. This tag guarantees that you use a specific version of Grafana instead of the most recent commit.

Reviewer: ```suggestion
To ensure stability and consistency, we strongly recommend using the `grafana/grafana-oss-dev:-pre` tag when running the Grafana main branch in a production environment. This tag ensures that you are using a specific version of Grafana instead of the most recent commit, which could potentially introduce new bugs or issues.
```

Author: This section also needs to be rechecked by a developer as I cannot find "dev" tag on the docker hub. Only "develop" i.e.
https://hub.docker.com/layers/grafana/grafana-oss/develop/images/sha256-9fb294499601d06fe643f1a8203cf46fd17c4e76e28869a755b6119231a314e4?context=explore

Reviewer 2: dev isn't a tag, there is the grafana-oss image and a separate grafana-oss-dev image (and the same arrangement for enterprise). This avoids polluting the tag namespace for the main grafana images with thousands of pre-release tags

For a list of available tags, refer to [grafana/grafana-oss](https://hub.docker.com/r/grafana/grafana-oss/tags/) and [grafana/grafana-oss-dev](https://hub.docker.com/r/grafana/grafana-oss-dev/tags/).

> # Default paths

Reviewer: ```suggestion
## Default paths
```

Reviewer: There can only be one H1 in a topic, which is the topic title.

> # Default paths

Reviewer: Can you please provide an introduction here so that I understand why we're talking about default paths. Just one or two sentences.

Author: done please check

> The following settings are hard-coded when launching the Grafana Docker container and can only be overridden using [environment variables](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#override-configuration-with-environment-variables), not in `conf/grafana.ini`.

Reviewer: ```suggestion
When starting the Grafana Docker container, the following configurations are set by default and cannot be changed by editing the `conf/grafana.ini` file. Instead, you can only modify these configurations using [environment variables]({{< relref "./configure-grafana/#override-configuration-with-environment-variables" >}}).
```

  - configuration
  - documentation
  - docker
  - docker compose
title: Configure a Grafana Docker image
menuTitle: Configure a Docker image
weight: 1800
---

# Configure a Grafana Docker image

Running Grafana on Docker may require some additional configuration which depends on the complexity of your environment. We will explain some of those e.g. using different images, changing logging levels, defining secrets on Cloud (e.g. AWS), configuring plugins etc.

> **Note:** The examples in this topic use the Grafana Enterprise Docker image. You can use the Grafana Open Source edition by changing the Docker image to `grafana/grafana-oss`.

# Supported Docker image variants

You can install and run Grafana using the following official Docker images.

- **Grafana Enterprise**: `grafana/grafana-enterprise`

- **Grafana Open Source**: `grafana/grafana-oss`

Each edition is available in two variants: Alpine and Ubuntu.

## Alpine image (recommended)

> [Alpine Linux](https://alpinelinux.org/about/) is an independent, non-commercial, general purpose Linux distribution designed for users users who appreciate security, simplicity and resource efficiency. Alpine Linux is much smaller than other distribution base images, allowing for slimmer and more secure images to be created.

Reviewer: ```suggestion
[Alpine Linux](https://alpinelinux.org/about/) is an independent, non-commercial, general-purpose Linux distribution designed for users who appreciate security, simplicity, and resource efficiency. Alpine Linux is much smaller than other distribution base images, allowing for slimmer and more secure images to be created.
```

By default, the images are built using the widely-used [Alpine Linux project](http://alpinelinux.org/) and can be accessed through the [Alpine official image](https://hub.docker.com/_/alpine).
If you prioritize security and want to minimize the size of your image, it is recommended that you use the Alpine variant. However, it's important to note that the Alpine variant uses [musl libc](http://www.musl-libc.org/) instead of [glibc and others](http://www.etalabs.net/compare_libcs.html). As a result, some software might encounter problems depending on their libc requirements. Nonetheless, most software should not experience any issues, so the Alpine variant is generally reliable.

## Ubuntu image

- **Grafana Enterprise**: `grafana/grafana-enterprise:<version>-ubuntu`

- **Grafana Open Source**: `grafana/grafana-oss:<version>-ubuntu`

These images are based on [Ubuntu](https://ubuntu.com/), available in the [Ubuntu official image](https://hub.docker.com/_/ubuntu). It is an alternative image for those who prefer an Ubuntu-based image and/or are dependent on certain tooling not available for Alpine.

## Run a specific version of Grafana

> You can also run a specific version, or run a beta version based on the main branch of the [grafana/grafana GitHub repository](https://github.com/grafana/grafana).

Reviewer: ```suggestion
You can also run a specific version of Grafana or a beta version based on the main branch of the [grafana/grafana GitHub repository](https://github.com/grafana/grafana).
```

The following command runs the Grafana Enterprise container, which has version 9.4.7. If you want to run a different version, modify the version number section.

```bash
docker run -d -p 3000:3000 --name grafana grafana/grafana-enterprise:9.4.7
```

## Run the Grafana main branch

For every successful build of the main branch, we update the `grafana/grafana-oss:main` and `grafana/grafana-oss:main-ubuntu` tags. Additionally, two new tags are created, `grafana/grafana-oss-dev:<version>-<build ID>pre` and `grafana/grafana-oss-dev:<version>-<build ID>pre-ubuntu`, where version is the next version of Grafana and build ID is the ID of the corresponding CI build. You can use these Tags to get access to the latest main builds of Grafana.
When running Grafana main in production, we strongly recommend that you use the `grafana/grafana-oss-dev:<version>-<build ID>pre` tag. This tag guarantees that you use a specific version of Grafana instead of the most recent commit.
For a list of available tags, refer to [grafana/grafana-oss](https://hub.docker.com/r/grafana/grafana-oss/tags/) and [grafana/grafana-oss-dev](https://hub.docker.com/r/grafana/grafana-oss-dev/tags/).

## Default paths

> Grafana have some predefined default configuraiton parameters which are always same when a fresh Grafana server is installed and is independent of the operating systems or the environment (e.g. virtual machine, docker, kubernetes etc.). You can refer the [Configure Grafana](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana) documentation to view all the default configuraiton settings.

Reviewer: ```suggestion
Grafana comes with default configuration parameters that remain the same among versions regardless of the operating system or the environment (for example, virtual machine, Docker, Kubernetes, etc.). You can refer to the [Configure Grafana](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana) documentation to view all the default configuraiton settings.
```

> When starting the Grafana Docker container, the following configurations are set by default and cannot be changed by editing the `conf/grafana.ini` file. Instead, you can only modify these configurations using [environment variables]({{< relref "./configure-grafana/#override-configuration-with-environment-variables" >}}).

Reviewer: ```suggestion
The following configurations are set by default when you start the Grafana Docker container. You cannot change the configurations by editing the `conf/grafana.ini` file. Instead, you can only modify these configurations using [environment variables]({{< relref "./configure-grafana/#override-configuration-with-environment-variables" >}}).
```

  - docker compose
title: Configure a Grafana Docker image
menuTitle: Configure a Docker image
weight: 1800
---

# Configure a Grafana Docker image

This topic explains how to run Grafana on Docker in complex environments that require you to:

- Use different images
- Change logging levels
- Define secrets on the Cloud
- Configure plugins

> **Note:** The examples in this topic use the Grafana Enterprise Docker image. You can use the Grafana Open Source edition by changing the Docker image to `grafana/grafana-oss`.

# Supported Docker image variants

You can install and run Grafana using the following official Docker images.

- **Grafana Enterprise**: `grafana/grafana-enterprise`

- **Grafana Open Source**: `grafana/grafana-oss`

Each edition is available in two variants: Alpine and Ubuntu.

## Alpine image (recommended)

[Alpine Linux](https://alpinelinux.org/about/) is a Linux distribution not affiliated with any commercial entity. It is a versatile operating system that caters to users who prioritize security, efficiency, and user-friendliness. Alpine Linux is much smaller than other distribution base images, allowing for slimmer and more secure images to be created.

> By default, the images are built using the widely-used [Alpine Linux project](http://alpinelinux.org/) and can be accessed through the [Alpine official image](https://hub.docker.com/_/alpine).

Reviewer 2: ```suggestion
By default, the images are built using the widely-used [Alpine Linux project](http://alpinelinux.org/) base image, which can be found in the [Alpine docker repo](https://hub.docker.com/_/alpine).
```

If you prioritize security and want to minimize the size of your image, it is recommended that you use the Alpine variant. However, it's important to note that the Alpine variant uses [musl libc](http://www.musl-libc.org/) instead of [glibc and others](http://www.etalabs.net/compare_libcs.html). As a result, some software might encounter problems depending on their libc requirements. Nonetheless, most software should not experience any issues, so the Alpine variant is generally reliable.

## Ubuntu image

> The Ubuntu image is maintained by [Canonical](https://launchpad.net/cloud-images). The Grafana Enterprise and OSS images are based on [Ubuntu](https://ubuntu.com/) and can be accessed through the [Ubuntu official image](https://hub.docker.com/_/ubuntu). This is a good option for users who prefer an Ubuntu-based image or require certain tools unavailable on Alpine.

Reviewer 2: ```suggestion
The Ubuntu-based Grafana Enterprise and OSS images are built using the [Ubuntu](https://ubuntu.com/) base image, which can be found in the [Ubuntu docker repo](https://hub.docker.com/_/ubuntu). An Ubuntu-based image can be a good option for users who prefer an Ubuntu-based image or require certain tools unavailable on Alpine.
```

When running Grafana main in production, we strongly recommend that you use the `grafana/grafana-oss-dev:<version>-<build ID>pre` tag. This tag guarantees that you use a specific version of Grafana instead of the most recent commit.
For a list of available tags, refer to [grafana/grafana-oss](https://hub.docker.com/r/grafana/grafana-oss/tags/) and [grafana/grafana-oss-dev](https://hub.docker.com/r/grafana/grafana-oss-dev/tags/).

## Default paths

Grafana comes with default configuration parameters that remain the same among versions regardless of the operating system or the environment (for example, virtual machine, Docker, Kubernetes, etc.). You can refer to the [Configure Grafana](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana) documentation to view all the default configuraiton settings.

> The following configurations are set by default when you start the Grafana Docker container. You cannot change the configurations by editing the `conf/grafana.ini` file. Instead, you can only modify these configurations using [environment variables]({{< relref "./configure-grafana/#override-configuration-with-environment-variables" >}}).

Reviewer 2: ```suggestion
The following configurations are set by default when you start the Grafana Docker container. When running in Docker you cannot change the configurations by editing the `conf/grafana.ini` file. Instead, you can modify the configuration using [environment variables]({{< relref "./configure-grafana/#override-configuration-with-environment-variables" >}}).
```

Each edition is available in two variants: Alpine and Ubuntu.

## Alpine image (recommended)

[Alpine Linux](https://alpinelinux.org/about/) is a Linux distribution not affiliated with any commercial entity. It is a versatile operating system that caters to users who prioritize security, efficiency, and user-friendliness. Alpine Linux is much smaller than other distribution base images, allowing for slimmer and more secure images to be created.

By default, the images are built using the widely-used [Alpine Linux project](http://alpinelinux.org/) and can be accessed through the [Alpine official image](https://hub.docker.com/_/alpine).
If you prioritize security and want to minimize the size of your image, it is recommended that you use the Alpine variant. However, it's important to note that the Alpine variant uses [musl libc](http://www.musl-libc.org/) instead of [glibc and others](http://www.etalabs.net/compare_libcs.html). As a result, some software might encounter problems depending on their libc requirements. Nonetheless, most software should not experience any issues, so the Alpine variant is generally reliable.

## Ubuntu image

> The Ubuntu image are maintained by [Canonical](https://launchpad.net/cloud-images). The Grafana Enterprise and OSS images are based on [Ubuntu](https://ubuntu.com/) and can be accessed through the [Ubuntu official image](https://hub.docker.com/_/ubuntu). This is a good option for users who prefer an Ubuntu-based image or require specific tools unavailable on Alpine.

Reviewer: ```suggestion
The Ubuntu image is maintained by [Canonical](https://launchpad.net/cloud-images). The Grafana Enterprise and OSS images are based on [Ubuntu](https://ubuntu.com/) and can be accessed through the [Ubuntu official image](https://hub.docker.com/_/ubuntu). This is a good option for users who prefer an Ubuntu-based image or require certain tools unavailable on Alpine.
```

To run a specific version of Grafana, add it in the command <version number> section:

```bash
docker run -d -p 3000:3000 --name grafana grafana/grafana-enterprise:<version number>
```

Example:

The following command runs the Grafana Enterprise container and specifies version 9.4.7. If you want to run a different version, modify the version number section.

```bash
docker run -d -p 3000:3000 --name grafana grafana/grafana-enterprise:9.4.7
```

## Run the Grafana main branch

After every successful build of the main branch, two tags, `grafana/grafana-oss:main` and `grafana/grafana-oss:main-ubuntu`, are updated. Additionally, two new tags are created: `grafana/grafana-oss-dev:-pre` and `grafana/grafana-oss-dev:-pre-ubuntu`, where version is the next version of Grafana and build ID is the ID of the corresponding CI build. These tags provide access to the most recent Grafana main builds.

> To ensure stability and consistency, we strongly recommend using the `grafana/grafana-oss-dev:-pre` tag when running the Grafana main branch in a production environment. This tag ensures that you are using a specific version of Grafana instead of the most recent commit, which could potentially introduce bugs or issues. It also avoids polluting the tag namespace for the main Grafana images with thousands of pre-release tags.

Reviewer 2: You've lost the placeholders for version and build id here so the tags don't make sense. The correct format should be something like `grafana/grafana-oss-dev:<version><build ID>-pre`

You can see the tags here https://hub.docker.com/r/grafana/grafana-oss-dev/tags (and we should add that url as a link in this doc so users can too).

Author: Fixed the path to the dev images and also added the reference link

| GF_PATHS_PLUGINS      | /var/lib/grafana/plugins  |
| GF_PATHS_PROVISIONING | /etc/grafana/provisioning |

> # Install plugins in the Docker container

Reviewer: ```suggestion
## Install plugins in the Docker container
```

> Just like users can install the publicly available plugins in Grafana, can do same for plugins that are private or used internally in an organisation.

Reviewer: ```suggestion
You can install the publicly available plugins in Grafana and can do the same for plugins that are private or used internally in an organization.
```

> ## Install plugins from other sources

Reviewer: ```suggestion
### Install plugins from other sources
```

> In order to install plugins from other sources, You will need to define the custom URL and specify it using the Environment Variable: `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`.

Reviewer: ```suggestion
To install plugins from other sources, you must define the custom URL and specify it in an environment variable: `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`.
```

Example:

> The following command will run Grafana Enterprise on **port 3000** in detached mode and also installs the custom plugin which is specified as a URL parameter under the `GF_INSTALL_PLUGIN` Environment variable.

Reviewer: ```suggestion
The following command runs Grafana Enterprise on **port 3000** in detached mode and installs the custom plugin, which is specified as a URL parameter in the `GF_INSTALL_PLUGIN` environment variable.
```

```bash
docker run -d \
  -p 3000:3000 \
  --name=grafana \
  -e "GF_INSTALL_PLUGINS=http://plugin-domain.com/my-custom-plugin.zip;custom-plugin,grafana-clock-panel" \
  grafana/grafana-enterprise
```

> # Build and run a Docker image with pre-installed plugins

Reviewer: ```suggestion
## Build and run a Docker image with pre-installed plugins
```

> You can build your own customized image that includes plugins. This saves time if you are creating multiple images and you want them all to have the same plugins installed on the build.

Reviewer: ```suggestion
If you need to create multiple images with the same plugins, you can save time by building your own customized image that includes those plugins.
```

In the Grafana GitHub repository, the `packaging/docker/custom/` folder includes a `Dockerfile` that you can use to build a custom Grafana image. The `Dockerfile` accepts `GRAFANA_VERSION`, `GF_INSTALL_PLUGINS`, and `GF_INSTALL_IMAGE_RENDERER_PLUGIN` as build arguments.

The `GRAFANA_VERSION` build argument must be a valid `grafana/grafana` Docker image tag. By default, Grafana builds an Alpine-based image. To build an Ubuntu-based image, append `-ubuntu` to the `GRAFANA_VERSION` build argument.

Example:

The following example shows you how to build and run a custom Grafana Docker image based on the latest official Ubuntu-based Grafana Docker image:

```bash
# go to the custom directory
cd packaging/docker/custom

# run the docker build command to build the image
docker build \
  --build-arg "GRAFANA_VERSION=latest-ubuntu" \
  -t grafana-custom .

# run the custom grafana container using docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

> ## Build with pre-installed plugins

Reviewer: Build what?

> ## Build with pre-installed plugins

Author: image i.e. Grafana Docker image

> ## Build with pre-installed plugins

Reviewer: Got it. thanks.

> You can build the Grafana custom image by including plugins which are available on the [Grafana Plugin download page](https://grafana.com/grafana/plugins). By this, you will not need to install plugins manually which can save time.

Reviewer: ```suggestion
To save time, you can customize a Grafana image by including plugins available on the [Grafana Plugin download page](https://grafana.com/grafana/plugins). By doing so, you won't have to manually install the plugins each time, making the process more efficient.
```

> > **Note:**  If you need to specify the version of a plugin, you can add it to the `GF_INSTALL_PLUGINS` build an argument. Otherwise, the latest will be assumed for e.g. `--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel 1.0.1,grafana-simple-json-datasource 1.3.5"`

Reviewer: ```suggestion
> **Note:** To specify the version of a plugin, you can use the `GF_INSTALL_PLUGINS` build argument and add the version number. The latest version is used if you don't specify a version number. For example, you can use `--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel 1.0.1,grafana-simple-json-datasource 1.3.5"` to specify the versions of two plugins.
```

Example:

The following explains as how to build a custom Grafana image and run it with plugins as pre-installed

```bash
# go to the custom directory
cd packaging/docker/custom

# running the build command
# include the plugins you want e.g. clock planel etc
docker build \
  --build-arg "GRAFANA_VERSION=latest" \
  --build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel,grafana-simple-json-datasource" \
  -t grafana-custom .

# running the custom Grafana container using the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

> ## Build with pre-installed plugins from other sources

Reviewer: Build what?

Author: image i.e. Grafana Docker image

Reviewer: Got it, thanks.

> You can build a Docker image with plugins from other sources if they are not publicly available (e.g. a private plugin used only within an organization) by specifying the URL like this: `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`.

Reviewer: ```suggestion
If a plugin is not publicly available and is only used within your organization, you can still build a Docker image with it by specifying the plugin's URL using the `GF_INSTALL_PLUGINS` build argument. You can do this by adding the URL and the plugin install folder name like this: `GF_INSTALL_PLUGINS=;`.
```

> Example:
> 
> 

Reviewer: ```suggestion
```

> The below example shows how to build a customized Grafana Docker image with the **clock panel** and also the **simple-json-datasource** plugin gets installed by defining them in the build-argument under Grafana Plugin environment variable 

Reviewer: ```suggestion
The following example demonstrates creating a customized Grafana Docker image that includes the clock panel and simple-json-datasource plugins. You can define these plugins in the build argument using the Grafana Plugin environment variable
```

```bash
#go to the folder
cd packaging/docker/custom

#running the build command
docker build \
--build-arg "GRAFANA_VERSION=latest" \
--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel, grafana-simple-json-datasource" \
-t grafana-custom .

# running the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

## Build Grafana with the Image Renderer plugin pre-installed

> > **Note:**  This is experimental.

Reviewer: ```suggestion
> **Note:**  This feature is experimental.
```

> The Grafana Image Renderer plugin does not currently work (see [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301)) if it is installed in a Grafana Docker image. You can build a custom Docker image by using the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument. This installs additional dependencies needed for the Grafana Image Renderer plugin to run.

Reviewer: ```suggestion
The Grafana Image Renderer plugin does not currently work (see [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301)) if it is installed in a Grafana Docker image. You can build a custom Docker image using the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument. This installs additional dependencies required for the Grafana Image Renderer plugin to run.
```

Example:

> The below example shows how to build the customized Grafana Docker Image and provides the Image Render Plugin in the build argument.

Reviewer: ```suggestion
The following example shows how to build a customized Grafana Docker image that includes the Image Renderer plugin.
```

```bash
#go to the folder
cd packaging/docker/custom

#running the build command
docker build \
--build-arg "GRAFANA_VERSION=latest" \
--build-arg "GF_INSTALL_IMAGE_RENDERER_PLUGIN=true" \
-t grafana-custom .

# running the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

# Logging

> Logs in the Docker container go to `STDOUT` by default, as is common in the Docker world. Change this by setting a different [log mode](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#mode). Available options are `console`, `file`, and `syslog`. The default is `console` and `file`. Use spaces to separate multiple modes, e.g `console file`

Reviewer: ```suggestion
By default, Docker container logs are directed to `STDOUT`, which is a common practice in the Docker community. You can change this by setting a different [log mode]({{< relref "../../../configure-grafana/#mode" >}}) such as `console`, `file`, or `syslog`. You can choose to use one or more modes by separating them with spaces, for example, `console file`. By default, both `console` and `file` modes are enabled.
```

Example:

To change the log mode, use the `GF_LOG_MODE` environment variable and define your mode e.g. to use `console file`, run the command:

```bash
# Run Grafana while logging to both standard out 
# and /var/log/grafana/grafana.log

docker run -p 3000:3000 -e "GF_LOG_MODE=console file" grafana/grafana-enterprise
```
# Configure Grafana with Docker Secrets

> You can supply Grafana sensitive information e.g. secrets, login credentials etc. with configuration through files. This works well with [Docker Secrets](https://docs.docker.com/engine/swarm/secrets/) as the secrets by default are mapped into the following location in the container: `/run/secrets/<name of secret>.
> ` 
> You can do this with any of the configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the path of the file holding the secret.

Reviewer: ```suggestion
You can supply Grafana sensitive information e.g. secrets, login credentials etc. with configuration through files. This works well with [Docker Secrets](https://docs.docker.com/engine/swarm/secrets/) as the secrets by default are mapped into the following location in the container: `/run/secrets/<name of secret>.
` 
You can do this with any of the configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the path of the file holding the secret.

You can provide sensitive information such as secrets and login credentials to Grafana by using configuration files. This method works particularly well with Docker Secrets, as the secrets are automatically mapped to the `/run/secrets/` location within the container. You can apply this technique to any of the configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the path of the file that contains the secret information.

```

> For example, you could set the admin password this way:

Reviewer: ```suggestion
The following example demonstrates how to set the admin password:
```

- Admin password secret: `/run/secrets/admin_password`
- Environment variable: `GF_SECURITY_ADMIN_PASSWORD__FILE=/run/secrets/admin_password`

> # Configure AWS credentials for CloudWatch Support

Reviewer: ```suggestion

## Configure AWS credentials for CloudWatch Support
```

> Defining secrets is very important on the public cloud so that it is not visible as plain text and provide a layer of security.

Reviewer: I'm not sure what this means. Can you pls clarify?

Author: rewrote this section in more better way. please check

> ```bash

Reviewer: none of these example commands have any context. Can you pls add some information about what's going on here?

Author: rewrote this section in more better way. please check

Reviewer: thank you. reviewing!

| GF_PATHS_PLUGINS      | /var/lib/grafana/plugins  |
| GF_PATHS_PROVISIONING | /etc/grafana/provisioning |

## Install plugins in the Docker container

> You can install publicly available plugins and also plugins that are private or used internally in an organization.

Reviewer: ```suggestion
You can install publicly available plugins and plugins that are private or used internally in an organization.
```

### Install plugins from other sources

To install plugins from other sources, you must define the custom URL and specify it in an environment variable: `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`.

Example:

The following command runs Grafana Enterprise on **port 3000** in detached mode and installs the custom plugin, which is specified as a URL parameter in the `GF_INSTALL_PLUGIN` environment variable.

```bash
docker run -d \
  -p 3000:3000 \
  --name=grafana \
  -e "GF_INSTALL_PLUGINS=http://plugin-domain.com/my-custom-plugin.zip;custom-plugin,grafana-clock-panel" \
  grafana/grafana-enterprise
```

## Build and run a Docker image with pre-installed plugins

> If you need to create multiple images with the same plugins, you can save time by building your own customized image that includes those plugins.

Reviewer: ```suggestion
To create multiple images with the same plugins, you can save time by building your own customized image that includes those plugins.
```

In the Grafana GitHub repository, the `packaging/docker/custom/` folder includes a `Dockerfile` that you can use to build a custom Grafana image. The `Dockerfile` accepts `GRAFANA_VERSION`, `GF_INSTALL_PLUGINS`, and `GF_INSTALL_IMAGE_RENDERER_PLUGIN` as build arguments.

The `GRAFANA_VERSION` build argument must be a valid `grafana/grafana` Docker image tag. By default, Grafana builds an Alpine-based image. To build an Ubuntu-based image, append `-ubuntu` to the `GRAFANA_VERSION` build argument.

Example:

The following example shows you how to build and run a custom Grafana Docker image based on the latest official Ubuntu-based Grafana Docker image:

```bash
# go to the custom directory
cd packaging/docker/custom

# run the docker build command to build the image
docker build \
  --build-arg "GRAFANA_VERSION=latest-ubuntu" \
  -t grafana-custom .

# run the custom grafana container using docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

> ## Build with pre-installed plugins

Reviewer: ```suggestion
## Build a Grafana Docker image with pre-installed plugins
```

To save time, you can customize a Grafana image by including plugins available on the [Grafana Plugin download page](https://grafana.com/grafana/plugins). By doing so, you won't have to manually install the plugins each time, making the process more efficient.

> **Note:** To specify the version of a plugin, you can use the `GF_INSTALL_PLUGINS` build argument and add the version number. The latest version is used if you don't specify a version number. For example, you can use `--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel 1.0.1,grafana-simple-json-datasource 1.3.5"` to specify the versions of two plugins.

Example:

> The following explains as how to build a custom Grafana image and run it with plugins as pre-installed

Reviewer: ```suggestion
The following example shows how to build and run a custom Grafana image with pre-installed plugins.
```

```bash
# go to the custom directory
cd packaging/docker/custom

# running the build command
# include the plugins you want e.g. clock planel etc
docker build \
  --build-arg "GRAFANA_VERSION=latest" \
  --build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel,grafana-simple-json-datasource" \
  -t grafana-custom .

# running the custom Grafana container using the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

> ## Build with pre-installed plugins from other sources

Reviewer: ```suggestion
## Build a Grafana Docker image with pre-installed plugins from other sources
```

> If a plugin is not publicly available and is only used within your organization, you can still build a Docker image with it by specifying the plugin's URL using the `GF_INSTALL_PLUGINS` build argument. You can do this by adding the URL and the plugin install folder name like this: `GF_INSTALL_PLUGINS=;`.

Reviewer: ```suggestion
You can create a Docker image containing a plugin that is exclusive to your organization, even if it is not accessible to the public. Simply use the `GF_INSTALL_PLUGINS` build argument to specify the plugin's URL and installation folder name, such as `GF_INSTALL_PLUGINS=;`.
```

> The following example demonstrates creating a customized Grafana Docker image that includes the clock panel and simple-json-datasource plugins. You can define these plugins in the build argument using the Grafana Plugin environment variable

Reviewer: ```suggestion
The following example demonstrates creating a customized Grafana Docker image that includes the clock panel and simple-json-datasource plugins. You can define these plugins in the build argument using the Grafana Plugin environment variable.
```

```bash
#go to the folder
cd packaging/docker/custom

#running the build command
docker build \
--build-arg "GRAFANA_VERSION=latest" \
--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel, grafana-simple-json-datasource" \
-t grafana-custom .

# running the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

## Build Grafana with the Image Renderer plugin pre-installed

> **Note:**  This feature is experimental.

> The Grafana Image Renderer plugin does not currently work (see [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301)) if it is installed in a Grafana Docker image. You can build a custom Docker image using the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument. This installs additional dependencies needed for the Grafana Image Renderer plugin to run.

Reviewer: ```suggestion
Currently, the Grafana Image Renderer plugin is not functional (as stated in [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301)) when installed in a Grafana Docker image. However, you can create a customized Docker image utilizing the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument as a solution. This will install the necessary dependencies for the Grafana Image Renderer plugin to run.
```

Example:

The following example runs Grafana using the `console file` log mode that is set in the `GF_LOG_MODE` environment variable

```bash
# Run Grafana while logging to both standard out 
# and /var/log/grafana/grafana.log

docker run -p 3000:3000 -e "GF_LOG_MODE=console file" grafana/grafana-enterprise
```
# Configure Grafana with Docker Secrets

Using configuration files, you can provide sensitive information such as secrets and login credentials to Grafana. This method works particularly well with Docker Secrets, as the secrets are automatically mapped to the `/run/secrets/` location within the container. 

You can apply this technique to any configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the file path that contains the secret information.

The following example demonstrates how to set the admin password:

- Admin password secret: `/run/secrets/admin_password`
- Environment variable: `GF_SECURITY_ADMIN_PASSWORD__FILE=/run/secrets/admin_password`

## Configure Docker secrets credentials for AWS CloudWatch

> Grafana ships with built-in support for [Amazon CloudWatch datasource](https://grafana.com/docs/grafana/latest/datasources/aws-cloudwatch/), where need to provide information such as AWS ID-Key, secret access key, region etc. For which you can use the Docker secrets to achieve it.

Reviewer: ```suggestion
Grafana ships with built-in support for the [Amazon CloudWatch datasource]({{< relref "../datasources/aws-cloudwatch/" >}}). To configure the data source, you must provide information such as the AWS ID-Key, secret access key, region, and so on. You can use Docker secrets as a way to provide this information.
```

Reviewer: I added a suggestion, but don't really understand it ;) - We are telling the user that they must provide the information and that Docker secrets can also provide the information. Can you pls help clarify?

| GF_PATHS_PLUGINS      | /var/lib/grafana/plugins  |
| GF_PATHS_PROVISIONING | /etc/grafana/provisioning |

## Install plugins in the Docker container

> You can install publicly available plugins and plugins that are private or used internally in an organization.

Reviewer 2: This should link back to the standard install instructions for plugins

### Install plugins from other sources

> To install plugins from other sources, you must define the custom URL and specify it in an environment variable: `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`.

Reviewer 2: ```suggestion
To install plugins from other sources, you must define the custom URL and specify it immediately before the plugin name in the `GF_INSTALL_PLUGINS` environment variable: `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin name>`.
```

Example:

> The following command runs Grafana Enterprise on **port 3000** in detached mode and installs the custom plugin, which is specified as a URL parameter in the `GF_INSTALL_PLUGIN` environment variable.

Reviewer 2: ```suggestion
The following command runs Grafana Enterprise on **port 3000** in detached mode and installs the custom plugin, which is specified as a URL parameter in the `GF_INSTALL_PLUGINS` environment variable.
```

```bash
docker run -d
  -p 3000:3000
  --name=grafana
  -e "GF_INSTALL_PLUGINS=http://plugin-domain.com/my-custom-plugin.zip;custom-plugin,grafana-clock-panel"
  grafana/grafana-enterprise
```

## Build and run a Docker image with pre-installed plugins

To create multiple images with the same plugins, you can save time by building your own customized image that includes those plugins.

In the Grafana GitHub repository, the `packaging/docker/custom/` folder includes a `Dockerfile` that you can use to build a custom Grafana image. The `Dockerfile` accepts `GRAFANA_VERSION`, `GF_INSTALL_PLUGINS`, and `GF_INSTALL_IMAGE_RENDERER_PLUGIN` as build arguments.

The `GRAFANA_VERSION` build argument must be a valid `grafana/grafana` Docker image tag. By default, Grafana builds an Alpine-based image. To build an Ubuntu-based image, append `-ubuntu` to the `GRAFANA_VERSION` build argument.

> Example:
> 
> The following example shows you how to build and run a custom Grafana Docker image based on the latest official Ubuntu-based Grafana Docker image:
> 
> ```bash
> # go to the custom directory
> cd packaging/docker/custom
> 
> # run the docker build command to build the image
> docker build
>   --build-arg "GRAFANA_VERSION=latest-ubuntu"
>   -t grafana-custom .
> 
> # run the custom grafana container using docker run command
> docker run -d -p 3000:3000 --name=grafana grafana-custom
> ```
> 
> ## Build a Grafana Docker image with pre-installed plugins
> 
> To save time, you can customize a Grafana image by including plugins available on the [Grafana Plugin download page](/grafana/plugins). By doing so, you won't have to manually install the plugins each time, making the process more efficient.
> 

Reviewer 2: This is all duplicative with little benefit, I suggest we collapse this section into 1
```suggestion
```

Reviewer: I spoke with @usmangt about this, and he's going to reach out to clarify what you want moved/combined.

Author: I have combined those 2 section and now availalbe as one under the heading "Build and run Grafana Docker image with pre-installed plugins"

> **Note:** To specify the version of a plugin, you can use the `GF_INSTALL_PLUGINS` build argument and add the version number. The latest version is used if you don't specify a version number. For example, you can use `--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel 1.0.1,grafana-simple-json-datasource 1.3.5"` to specify the versions of two plugins.

Example:

The following example shows how to build and run a custom Grafana image with pre-installed plugins.

```bash
# go to the custom directory
cd packaging/docker/custom

# running the build command
# include the plugins you want e.g. clock planel etc
docker build
  --build-arg "GRAFANA_VERSION=latest"
  --build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel,grafana-simple-json-datasource"
  -t grafana-custom .

# running the custom Grafana container using the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

> ## Build a Grafana Docker image with pre-installed plugins from other sources

Reviewer 2: This section doesn't actually show the user how to preinstall a plugin from another source, which requires prefixing the plugin slug with the url of a custom .zip file

You can create a Docker image containing a plugin that is exclusive to your organization, even if it is not accessible to the public. Simply use the `GF_INSTALL_PLUGINS` build argument to specify the plugin's URL and installation folder name, such as `GF_INSTALL_PLUGINS=;`.

The following example demonstrates creating a customized Grafana Docker image that includes the clock panel and simple-json-datasource plugins. You can define these plugins in the build argument using the Grafana Plugin environment variable.

```bash
#go to the folder
cd packaging/docker/custom

#running the build command
docker build
--build-arg "GRAFANA_VERSION=latest"
--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel, grafana-simple-json-datasource"
-t grafana-custom .

# running the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

> ## Build Grafana with the Image Renderer plugin pre-installed

Reviewer 2: This is likely worth moving up above the preinstalled plugins content, as it's something that does require using the custom image while plugins can be installed at boot time without needing a custom image.

Reviewer: Done.

> **Note:** This feature is experimental.

> Currently, the Grafana Image Renderer plugin is not functional (as stated in [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301)) when installed in a Grafana Docker image. However, you can create a customized Docker image utilizing the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument as a solution. This will install the necessary dependencies for the Grafana Image Renderer plugin to run.

Reviewer 2: ```suggestion
Currently, the Grafana Image Renderer plugin requires dependencies which are not available in the Grafana Docker image (see [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301) for more details). However, you can create a customized Docker image utilizing the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument as a solution. This will install the necessary dependencies for the Grafana Image Renderer plugin to run.
```

You can apply this technique to any configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the file path that contains the secret information. For more information about Docker secret command usage, refer to [docker secret](https://docs.docker.com/engine/reference/commandline/secret/).

The following example demonstrates how to set the admin password:

- Admin password secret: `/run/secrets/admin_password`
- Environment variable: `GF_SECURITY_ADMIN_PASSWORD__FILE=/run/secrets/admin_password`

### Configure Docker secrets credentials for AWS CloudWatch

Grafana ships with built-in support for the [Amazon CloudWatch datasource]({{< relref "../datasources/aws-cloudwatch" >}}). To configure the data source, you must provide information such as the AWS ID-Key, secret access key, region, and so on. You can use Docker secrets as a way to provide this information.

Example:

The following example defines how to use Grafana environment variables via Docker Secrets for the AWS ID-Key, secret access key, region, and profile.

```bash
docker run -d
-p 3000:3000
--name=grafana
-e "GF_AWS_PROFILES=default"
-e "GF_AWS_default_ACCESS_KEY_ID=YOUR_ACCESS_KEY"

> -e "GF_AWS_default_SECRET_ACCESS_KEY=YOUR_SECRET_KEY"

Reviewer 2: It may make sense to show how this would be done using a docker secret, rather than encouraging people to pass it via the command line.

-e "GF_AWS_default_REGION=us-east-1"
grafana/grafana-enterprise
```

You can also specify multiple profiles to `GF_AWS_PROFILES` (for example, `GF_AWS_PROFILES=default another`).

Supported variables:

- `GF_AWS_${profile}_ACCESS_KEY_ID`: AWS access key ID (required).
- `GF_AWS_${profile}_SECRET_ACCESS_KEY`: AWS secret access key (required).
- `GF_AWS_${profile}_REGION`: AWS region (optional).

> ## Troubleshoot a Docker deployment

Reviewer: @usmangt , please review this entire section as I moved some stuff around and re-worded a few things.

> The following is the list of the supported environment variables:

Reviewer: ```suggestion
The following list includes the supported environment variables:
```

By default, the Grafana log level is set to `INFO`, but you can increase the log level to `DEBUG` mode when you try to reproduce a problem.

For more information about logging, refer to [logs]({{< relref "./configure-grafana#log" >}}).

### Increase log level using the Docker run (CLI) command

To increase the log level to `DEBUG` mode, add the environment variable `GF_LOG_LEVEL` to the command line.

```bash
docker run -d -p 3000:3000 --name=grafana
-e "GF_LOG_LEVEL=DEBUG"
grafana/grafana-enterprise
```

### Increase log level using the Docker Compose

To increase the log level to `DEBUG` mode, add the environment variable `GF_LOG_LEVEL` to the `docker-compose.yaml` file.

```yaml
version: '3.8'
services:
  grafana:
    image: grafana/grafana-enterprise
    container_name: grafana
    restart: unless-stopped
    environment:
      # increases the log level from info to debug

>       - GF_LOG_LEVEL=debug

Reviewer 2: let's be consistent, either DEBUG or debug in both examples.

    ports:
      - '3000:3000'
    volumes:
      - 'grafana_storage:/var/lib/grafana'
volumes:
  grafana_storage: {}
```

### Validate Docker Compose YAML file

The chance syntax errors to appear in a YAML file increases as the file becomes more complex. You can use the following command to check for syntax errors.

```bash
docker compose config docker-compose.yaml
```

> The output of the command highlights errors in specific lines. Else it will output the content of the `docker-compose.yaml` file in detailed YAML format.

Reviewer: @usmangt , I'm don't understand the sentence that begins with "Else it...". can you please clarify what this means? Thanks!

> The chance syntax errors to appear in a YAML file increases as the file becomes more complex. You can use the following command to check for syntax errors.

Reviewer 2: ```suggestion
The chance of syntax errors appearing in a YAML file increases as the file becomes more complex. You can use the following command to check for syntax errors.
```

Example:

The following example runs Grafana using the `console file` log mode that is set in the `GF_LOG_MODE` environment variable.

```bash
# Run Grafana while logging to both standard out
# and /var/log/grafana/grafana.log

docker run -p 3000:3000 -e "GF_LOG_MODE=console file" grafana/grafana-enterprise
```

## Configure Grafana with Docker Secrets

Using configuration files, you can input confidential data like login credentials and secrets into Grafana. This method works well with [Docker Secrets](https://docs.docker.com/engine/swarm/secrets/), as the secrets are automatically mapped to the `/run/secrets/` location within the container.

You can apply this technique to any configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the file path that contains the secret information. For more information about Docker secret command usage, refer to [docker secret](https://docs.docker.com/engine/reference/commandline/secret/).

The following example demonstrates how to set the admin password:

- Admin password secret: `/run/secrets/admin_password`
- Environment variable: `GF_SECURITY_ADMIN_PASSWORD__FILE=/run/secrets/admin_password`

> ### Configure Docker secrets credentials for AWS CloudWatch

Author: This is now fixed!!

-t grafana-custom .

# running the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

## Logging

By default, Docker container logs are directed to `STDOUT`, a common practice in the Docker community. You can change this by setting a different [log mode]({{< relref "../../../configure-grafana/#mode" >}}) such as `console`, `file`, or `syslog`. You can use one or more modes by separating them with spaces, for example, `console file`. By default, both `console` and `file` modes are enabled.

Example:

The following example runs Grafana using the `console file` log mode that is set in the `GF_LOG_MODE` environment variable.

```bash
# Run Grafana while logging to both standard out 
# and /var/log/grafana/grafana.log

docker run -p 3000:3000 -e "GF_LOG_MODE=console file" grafana/grafana-enterprise
```
## Configure Grafana with Docker Secrets

Using configuration files, you can input confidential data like login credentials and secrets into Grafana. This method works well with [Docker Secrets](https://docs.docker.com/engine/swarm/secrets/), as the secrets are automatically mapped to the `/run/secrets/` location within the container. 

> You can apply this technique to any configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the file path that contains the secret information. Please check the official [docker secret command usage](https://docs.docker.com/engine/reference/commandline/secret/) for more refrence.

Reviewer: ```suggestion
You can apply this technique to any configuration options in `conf/grafana.ini` by setting `GF_<SectionName>_<KeyName>__FILE` to the file path that contains the secret information. For more information about Docker secret command usage, refer to [docker secret](https://docs.docker.com/engine/reference/commandline/secret/).
```

  grafana/grafana-enterprise
```

## Build a custom Grafana Docker image

In the Grafana GitHub repository, the `packaging/docker/custom/` folder includes a `Dockerfile` that you can use to build a custom Grafana image. The `Dockerfile` accepts `GRAFANA_VERSION`, `GF_INSTALL_PLUGINS`, and `GF_INSTALL_IMAGE_RENDERER_PLUGIN` as build arguments.

The `GRAFANA_VERSION` build argument must be a valid `grafana/grafana` Docker image tag. By default, Grafana builds an Alpine-based image. To build an Ubuntu-based image, append `-ubuntu` to the `GRAFANA_VERSION` build argument.

Example:

The following example shows you how to build and run a custom Grafana Docker image based on the latest official Ubuntu-based Grafana Docker image:

```bash
# go to the custom directory
cd packaging/docker/custom

# run the docker build command to build the image
docker build \
  --build-arg "GRAFANA_VERSION=latest-ubuntu" \
  -t grafana-custom .

# run the custom grafana container using docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

### Build Grafana with the Image Renderer plugin pre-installed

> **Note:** This feature is experimental.

> Currently, the Grafana Image Renderer plugin requires dependencies which are not available in the Grafana Docker image (see [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301) for more details). However, you can create a customized Docker image utilizing the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument as a solution. This will install the necessary dependencies for the Grafana Image Renderer plugin to run.

Reviewer: ```suggestion
Currently, the Grafana Image Renderer plugin requires dependencies that are not available in the Grafana Docker image (see [GitHub Issue#301](https://github.com/grafana/grafana-image-renderer/issues/301) for more details). However, you can create a customized Docker image utilizing the `GF_INSTALL_IMAGE_RENDERER_PLUGIN` build argument as a solution. This will install the necessary dependencies for the Grafana Image Renderer plugin to run.
```

> **Note:** To specify the version of a plugin, you can use the `GF_INSTALL_PLUGINS` build argument and add the version number. The latest version is used if you don't specify a version number. For example, you can use `--build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel 1.0.1,grafana-simple-json-datasource 1.3.5"` to specify the versions of two plugins.

Example:

The following example shows how to build and run a custom Grafana Docker image with pre-installed plugins.

```bash
# go to the custom directory
cd packaging/docker/custom

# running the build command
# include the plugins you want e.g. clock planel etc
docker build \
  --build-arg "GRAFANA_VERSION=latest" \
  --build-arg "GF_INSTALL_PLUGINS=grafana-clock-panel,grafana-simple-json-datasource" \
  -t grafana-custom .

# running the custom Grafana container using the docker run command
docker run -d -p 3000:3000 --name=grafana grafana-custom
```

### Build a Grafana Docker image with pre-installed plugins from other sources

> You can create a Docker image containing a plugin that is exclusive to your organization, even if it is not accessible to the public. Simply use the `GF_INSTALL_PLUGINS` build argument to specify the plugin's URL and installation folder name, such as `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`

Reviewer: ```suggestion
You can create a Docker image containing a plugin that is exclusive to your organization, even if it is not accessible to the public. Simply use the `GF_INSTALL_PLUGINS` build argument to specify the plugin's URL and installation folder name, such as `GF_INSTALL_PLUGINS=<url to plugin zip>;<plugin install folder name>`.
```

> The following example demonstrates creating a customized Grafana Docker image that includes a custom plugin from a URL link, the clock panel and simple-json-datasource plugins. You can define these plugins in the build argument using the Grafana Plugin environment variable.

Reviewer: ```suggestion
The following example demonstrates creating a customized Grafana Docker image that includes a custom plugin from a URL link, the clock panel plugin, and the simple-json-datasource plugin. You can define these plugins in the build argument using the Grafana Plugin environment variable.
```

The example uses the following values for the AWS Cloudwatch data source:

```bash
AWS_default_ACCESS_KEY_ID=aws01us02
AWS_default_SECRET_ACCESS_KEY=topsecret9b78c6
AWS_default_REGION=us-east-1
```

1. Create a Docker secret for each of the values noted above.

   ```bash
   echo "aws01us02" | docker secret create aws_access_key_id -
   ```

   ```bash
   echo "topsecret9b78c6" | docker secret create aws_secret_access_key -
   ```

   ```bash
   echo "us-east-1" | docker secret create aws_region -
   ```

1. Run the following command to determine that the secrets were created.

   ```bash
   $ docker secret ls
   ```

   The output from the command should look similar to the following:

   ```
   ID                          NAME           DRIVER    CREATED              UPDATED
   i4g62kyuy80lnti5d05oqzgwh   aws_access_key_id             5 minutes ago        5 minutes ago
   uegit5plcwodp57fxbqbnke7h   aws_secret_access_key         3 minutes ago        3 minutes ago
   fxbqbnke7hplcwodp57fuegit   aws_region                    About a minute ago   About a minute ago
   ```

   Where:

>    ID = the secret unique ID which we will use in the docker run command

Reviewer: ```suggestion
   ID = the secret unique ID that you will use in the docker run command
```

>    NAME = the we defined for each secret

Reviewer: @usmangt , not sure what this means. let's not use the word 'we'. Could this be "That you defined for each secret"?

Author: fixed now!!

1. Add the secrets to the command line when you run Docker.

   ```bash
   docker run -d -p 3000:3000 --name grafana \
     -e "GF_DEFAULT_INSTANCE_NAME=my-grafana" \
     -e "GF_AWS_PROFILES=default" \
     -e "GF_AWS_default_ACCESS_KEY_ID__FILE=/run/secrets/aws_access_key_id" \
     -e "GF_AWS_default_SECRET_ACCESS_KEY__FILE=/run/secrets/aws_secret_access_key" \
     -e "GF_AWS_default_REGION__FILE=/run/secrets/aws_region" \
     -v grafana-data:/var/lib/grafana \
     grafana/grafana-enterprise
   ```

You can also specify multiple profiles to `GF_AWS_PROFILES` (for example, `GF_AWS_PROFILES=default another`).

The following list includes the supported environment variables:

- `GF_AWS_${profile}_ACCESS_KEY_ID`: AWS access key ID (required).
- `GF_AWS_${profile}_SECRET_ACCESS_KEY`: AWS secret access key (required).
- `GF_AWS_${profile}_REGION`: AWS region (optional).

## Troubleshoot a Docker deployment

> By default, the Grafana log level is set to `INFO`, but you can increase the log level to `DEBUG` mode when you try to reproduce a problem.

Reviewer: ```suggestion
By default, the Grafana log level is set to `INFO`, but you can increase the log level to `DEBUG` mode when you want to reproduce a problem.
```

1. Run the following command to determine that the secrets were created.

   ```bash
   $ docker secret ls
   ```

   The output from the command should look similar to the following:

   ```
   ID                          NAME           DRIVER    CREATED              UPDATED
   i4g62kyuy80lnti5d05oqzgwh   aws_access_key_id             5 minutes ago        5 minutes ago
   uegit5plcwodp57fxbqbnke7h   aws_secret_access_key         3 minutes ago        3 minutes ago
   fxbqbnke7hplcwodp57fuegit   aws_region                    About a minute ago   About a minute ago
   ```

   Similarly, you can also find more information about the above created secrets using the `docker inspect` command.

   For example, to inspect the `aws_access_key_id` secret, run the following command:

   ```bash
   $ docker inspect aws_access_key_id
   [
       {
           "ID": "i4g62kyuy80lnti5d05oqzgwh",
           "Version": {
               "Index": 16
           },
           "CreatedAt": "2023-05-29T21:43:37.247825144Z",
           "UpdatedAt": "2023-05-29T21:43:37.247825144Z",
           "Spec": {
               "Name": "aws_access_key_id",
               "Labels": {}
           }
       }
   ]

>    ```

Reviewer: I don't understand the purpose of this section here. What do we mean by 'more information'? Is it important in the overall completion of this task? If not, I recommend removing this content.

Author: I have now removed this section

-e "GF_AWS_default_REGION=us-east-1" \
grafana/grafana-enterprise
```
You may also specify multiple profiles to `GF_AWS_PROFILES` (e.g. `GF_AWS_PROFILES=default another`).

Supported variables:

- `GF_AWS_${profile}_ACCESS_KEY_ID`: AWS access key ID (required).
- `GF_AWS_${profile}_SECRET_ACCESS_KEY`: AWS secret access key (required).
- `GF_AWS_${profile}_REGION`: AWS region (optional).

> # Troubleshooting

Reviewer: ```suggestion
## Troubleshooting
```

> For troubleshooting, we recommend increasing the log level to `DEBUG` mode. For more information, refer to the [log section in Configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#log).

Reviewer: I want to revisit this section after you've addressed the above suggestions.

-e "GF_AWS_default_REGION=us-east-1" \
grafana/grafana-enterprise
```

> You may also specify multiple profiles to `GF_AWS_PROFILES` (e.g. `GF_AWS_PROFILES=default another`).

Reviewer: ```suggestion
You can also specify multiple profiles to `GF_AWS_PROFILES` (for example, `GF_AWS_PROFILES=default another`).
```
